Expose a Factory Function

MikeyBurkman opened the pull request:
Alright, this is probably an odd proposal/PR, but I needed to expose Sync as a factory function for a very specific use case: load testing.

The best way we could think of load-testing our sync server realistically was to spin up a bazillion sync clients, to best reproduce a real-world scenario. Ideally we'd keep spinning up new sync clients until the server started choking, and that would give us a rough estimate on the number of simultaneous users we could support.

However, spinning up a browser instance for each of hundreds or thousands of users is not a great idea. So I thought to change Sync so we would spin up multiple clients in a single browser instance. This involved wrapping the global(-ish) variables that Sync uses in a the factory function's scope, and adding a "name" to the Lawnchair config, so the web storage databases are isolated. Now I can spin up a few dozen sync clients in each browser tab (depending on the sync frequency), and we can load test using just a few devices.

Now `fh.sync` is set to a default instance, with the Lawnchair name set to `undefined`, so this is a fully backwards-incompatible change. We simply expose a new `fh.syncFactory(id)` function as well. 

What do you think?

MikeyBurkman committed:
Include grunt-cli and build script

MikeyBurkman committed:
Sync factory function

MikeyBurkman committed:
Explicit build task

MikeyBurkman committed:
Fix gruntfile changes

MikeyBurkman commented on Gruntfile.js:
```diff
@@ -146,6 +146,7 @@ module.exports = function(grunt) {
 
   grunt.registerTask('concat-core-sdk', ['jshint',  'concat:lawnchair', 'concat:crypto', 'browserify:dist']);
 
+  grunt.registerTask('build', ['concat-core-sdk', 'uglify:dist']);
```
I added an explicit build task, because in our test client, I'm pointing directly at github, and hence I needed a command to build. (Running through all the tests wasn't always working when being built in the post-install script from another project.) This is not strictly needed for the factory function.

MikeyBurkman commented on package.json:
```diff
@@ -45,6 +46,7 @@
     "express": "3.4.8",
     "grunt": "0.4.2",
     "grunt-browserify": "1.3.2",
+    "grunt-cli": "1.2.0",
```
Adding this means that you don't need to have grunt globally installed. The npm scripts will use the local version of grunt to run the builds.

MikeyBurkman commented on src/sync-client.js:
```diff
@@ -5,1157 +5,1178 @@ var cidProvider = require('./clientIdProvider');
-    self.consoleLog('manage - START');
+module.exports = newClient;
+
+function newClient(id) {
```
99% of the changes to this file are indenting the extra level due to everything being wrapped in this `newClient(id)` function.

MikeyBurkman commented on src/sync-client.js:
```diff
@@ -5,1157 +5,1178 @@ var cidProvider = require('./clientIdProvider');
+    },
+
+    /**
+     * Sets clou
```
Github diffs are sometimes not very helpful... This line is the only other real change to this file. 

(Also, hooray for abstractions that make this kind of thing super easy to implement!)

MikeyBurkman commented on src/sync-client.js:
```diff
@@ -1166,159 +1187,149 @@ var self = {
+
+  self.setCloudHandler(defaultCloudHandler.handler);
+
+  return {
```
This is the only other change. Obviously should return this object instead of exporting, since we're now exporting the factory function.

wtrocki commented:
Just one quick comment from me. Make sure that each client have different cuid. If they are the same server will work with single client instance and you may get a lots of strange results.

MikeyBurkman commented:
@wtrocki Ah yes, I missed that. I guess I could prefix/postfix the cuid with the id passed to the factory function? Not sure if that's the best way but I'll try it out and see what happens.

MikeyBurkman committed:
Use new sync id in clientId

wei-lee commented on src/sync-client.js:
```diff
@@ -5,1157 +5,1180 @@ var cidProvider = require('./clientIdProvider');
-
-  manage: function(dataset_id, opts, query_params, meta_data, cb) {
-    self.consoleLog('manage - START');
+module.exports = newClient;
```
the tests are broken because they are referencing this file directly, so to fix it, we can either still expose the singleton in this file, and attach the factory function to it, or change the tests to create the instance first.

wei-lee commented:
@MikeyBurkman good idea and nice job, previously we have to spin up quite a few VMs to run the load tests, but this change will make it much easier. Thanks.

MikeyBurkman committed:
Fix unit tests and TS typings

MikeyBurkman committed:
minor typings fixes

wtrocki commented:
@MikeyBurkman Please let me know once you will be happy with your changes. I will run this on our automated test platform (it's to heavy to run in travis) and see what impact this has on fh-js-sdk.

@wei-lee Can I assume that you viewed and accepted proposal? 

ping @david-martin

MikeyBurkman commented:
@wtrocki Ah yeah, wanted to get approval from you all before moving forward. The latest changes:
1. Added the `factory` function on to `fh.sync` itself. Since `index.js` exports `fh.sync`, you wouldn't be able to see the `factory` function except through the global `fh.sync` variable, which didn't seem right.
2. The tests just call the syncClient factory function directly now. None of them test the `id` parameter, and to be honest I'm not immediately sure how we'd do automated tests for that functionality.
3. Fixed up the TS typings to include the `factory` function. I didn't seem to need the `module` declaration -- I was testing out the typings in a pet TS project, and it seemed to all compile fine. Still, I'm not the greatest at typings files, so if anyone else sees issues with my changes, let me know.

Are there any suggestions for a better name than `factory`? I was thinking maybe changing it to `build()` or `createNew()`.

Also, one further question. In the typings, `cloudUrl` is required in the SyncOptions interface. I looked through the code and it seems to be optional (the client would just use the same host if it weren't provided), but if it's optional, I'll fix that real quick.

MikeyBurkman committed:
travis fixes maybe

MikeyBurkman committed:
oops

MikeyBurkman committed:
ignore me squash these commits eventually

wei-lee commented:
yes, +1 from me.

MikeyBurkman commented on .travis.yml:
```diff
@@ -1,10 +1,9 @@
 before_install:
-  - npm install -g npm@2.13.5
-  - npm install -g grunt-cli
+  - npm install -g npm@3
```
We probably don't actually need this before_install block at all, now that I think about it...

wtrocki commented on fh-sync-js.d.ts:
```diff
@@ -2,237 +2,190 @@
 // Project: https://github.com/feedhenry/fh-sync-js
```
You will make hart time to merge @evanshortiss PR after: https://github.com/feedhenry/fh-sync/pull/41/files
I think it's best to not add this changes to this PR (it should be targeting single problem) and contribute type changes to PR above.

MikeyBurkman replied:
Strictly speaking, these changes *are* necessary though, as the factory function is not exposed without these changes.

Also, Evan's changes are for the cloud code, so I'm not sure how there will be a conflict?

wtrocki replied:
My bad. Sorry, got it mixed by names. Feel free to ignore this.

MikeyBurkman replied:
:) Evan and I were chatting about the typings last night actually, and we were getting mixed up ourselves. I guess there's no chance of changing the repo names to be something like `fh-sync-client` and `fh-sync-server`? :P

wtrocki replied:
fh-sync-client is not a way to go as we have multiple platforms. For example: https://github.com/feedhenry/fh-sync-android

MikeyBurkman replied:
`fh-sync-client-js`? The fact that fh-sync is also JS kind of indicates to me that we need a better qualifier for the JS client code than `-js`.

wtrocki replied:
That sounds ok. I will run that on mailing list.

MikeyBurkman replied:
Granted, that would create an inconsistency in naming unless you also renamed the client repos. `fh-sync-client-android`, etc. Renaming `fh-sync` to `fh-sync-server` would probably be sufficient.

wtrocki commented:
Tested in RainCatcher. 
@aidenkeating - are you ok with this change? 
This change breaks backwards compatibility.

wtrocki committed:
Version bump

aidenkeating commented:
@wtrocki Is this change backwards compatible? If not, should be be bumping a major version here? We'll also need to update the shrinkwrap?

Edit: Yep though, looks like a good idea, change is fine with me

wtrocki commented on test/tests/test_sync_manage.js:
```diff
@@ -1,6 +1,6 @@
 var process = require("process");
 
-var syncClient = require("../../src/sync-client");
+var syncClient = require("../../src/sync-client")();
```
I have done some research where is that used. This change will cause a little bit of headache as it will not be backwards compatible. I think this change is solid and we should have no issues with accepting it, however with this format it can cause problems with multiple sample apps, templates, performance tests, openshift starters etc.

wei-lee replied:
@wtrocki just curious where is the `sync-client.js` is being used directly, instead of using `index.js` file?

wtrocki replied:
Actually factory is there and everything works from the js side. I had problems with TypeScript but this was resolved already.

wtrocki commented:
**Thanks for contributing to Raincatcher**  :1st_place_medal: 
Your change was reviewed and accepted by our team

Changes published to npm
